pkg/traceql/engine_metrics.go

```diff
@@ -59,38 +59,50 @@ func IntervalCount(start, end, step uint64) int {
 	end = alignEnd(start, end, step)
 
 	intervals := (end - start) / step
-	intervals++
 	return int(intervals)
 }
 
 // TimestampOf the given interval with the start and step.
 func TimestampOf(interval, start, end, step uint64) uint64 {
 	start = alignStart(start, end, step)
-	return start + interval*step
+	// start as initial offset plus interval's offset
+	return start + (interval+1)*step
 }
 
 // IntervalOf the given timestamp within the range and step.
+// First interval is (start; start+step]
+// Last interval is (end-step; end]
+// The first interval's left border is limited to 0
 func IntervalOf(ts, start, end, step uint64) int {
 	if isInstant(start, end, step) { // always one interval
-		if !isTsValidForInterval(ts, start, end, step) {
+		if !isTsValidForInstant(ts, start, end) {
 			return -1
 		}
 		return 0
 	}
 
 	start = alignStart(start, end, step)
-	end = alignEnd(start, end, step) + step
+	end = alignEnd(start, end, step)
 
 	if !isTsValidForInterval(ts, start, end, step) {
 		return -1
 	}
+	if ts <= start { // to avoid overflow
+		return 0 // if pass validation and less than start, always first interval
+	}
 
-	return int((ts - start) / step)
+	interval := (ts - start - 1) / step
```

Comment:
Is there a clearer way to write this or add comments?   I find subtracting a single nanosecond confusing.  It's done this way because we are upper-bound inclusive.